File: src/main/java/ru/dkolmogortsev/task/storage/TimeEntriesStorage.java
package ru.dkolmogortsev.task.storage;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Function;
import java.util.stream.Collectors;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.infinispan.Cache;
import org.infinispan.query.Search;
import org.infinispan.query.dsl.QueryFactory;
import org.infinispan.query.dsl.SortOrder;
import ru.dkolmogortsev.task.TimeEntry;

/**
 * Created by dkolmogortsev on 2/25/17.
 */
@Singleton
public class TimeEntriesStorage
{
    Cache<Long, TimeEntry> infStorage;

    private QueryFactory factory;

    @Inject
    public TimeEntriesStorage(InfinispanCacheManager manager)
    {
        infStorage = manager.timeEntryStorage();
        factory = Search.getQueryFactory(infStorage);
    }

    public TimeEntry save(TimeEntry entry)
    {
        return infStorage.put(entry.getId(), entry);
    }

    public TimeEntry get(long timeEntryId)
    {
        return infStorage.get(timeEntryId);
    }

    public TimeEntry delete(long timeEntryId)
    {
        return infStorage.remove(timeEntryId);
    }

    public List<TimeEntry> getByEntryDate(long entryDate)
    {
        List<TimeEntry> entries = factory.from(TimeEntry.class).having("entryDate").eq(entryDate)
                .orderBy("start", SortOrder.ASC).build().list();
        return entries;
    }

    public Map<Long, List<TimeEntry>> getEntriesGroupedByDay()
    {
        List<Object[]> resultSet = factory.from(TimeEntry.class).select("entryDate").groupBy("entryDate")
                .orderBy("entryDate", SortOrder.ASC).build().list();
        Map<Long, List<TimeEntry>> longListMap = resultSet.stream().map(longs -> (long)longs[0]).collect(Collectors
                .toMap(Function.identity(), this::getByEntryDate, (entries, entries2) -> entries, TreeMap::new));
        return longListMap;

    }
}

File: src/main/kotlin/ru/dkolmogortsev/task/storage/TimeEntriesStorage.kt
package ru.dkolmogortsev.task.storage

import org.infinispan.Cache
import ru.dkolmogortsev.task.TimeEntry
import tornadofx.Component
import tornadofx.ScopedInstance

/**
 * Created by dkolmogortsev on 2/25/17.
 */
class TimeEntriesStorage : Component(), Storage<TimeEntry>, ScopedInstance {
    internal var infStorage: Cache<Long, TimeEntry> = InfinispanCacheManager.timeEntryStorage()
    override fun save(entry: TimeEntry) {
        infStorage.put(entry.id, entry)
    }

    override operator fun get(timeEntryId: Long): TimeEntry {
        return infStorage[timeEntryId]!!
    }

    fun delete(timeEntryId: Long): TimeEntry {
        return infStorage.remove(timeEntryId)!!
    }

    fun getByEntryDate(entryDate: Long): List<TimeEntry> {
        //TODO implement by lucene
        return listOf()
    }

    //TODO implement by lucene
    val entriesGroupedByDay: Map<Long, List<TimeEntry>>
        get() {
            return mapOf()
        }

    init {
//		factory = Search.getQueryFactory(infStorage)
//		val brokenEntries = factory.from(TimeEntry::class.java).having("end").equal(0).build().list<TimeEntry>()
//		brokenEntries.forEach { timeEntry ->
//			timeEntry.stop()
//			infStorage.put(timeEntry.id, timeEntry)
//		}
    }
}
